src/protocol/boolean/prefix_or.rs

```diff
@@ -0,0 +1,365 @@
+            .collect()
+    }
+
+    /// Step 5. `for i,j=1..λ, [g_(i,j)] = MULT([f_i], [a_(i,j)])`
+    ///
+    /// The interpretation is, `g_(i,j) = a_(i,j)` iff `f_i` = 1.
+    ///
+    /// # Example
+    /// ```ignore
+    ///   [a] = 0 0 0 0  0 0 1 0  0 1 0 1  0 0 0 0,   [f] = 0 1 0 0
+    ///   [g] = 0 0 0 0  0 0 1 0  0 0 0 0  0 0 0 0
+    /// ```
+    async fn step5(
+        f: &[Replicated<B>],
+        a: &[Replicated<B>],
+        ctx: ProtocolContext<'_, B>,
+    ) -> Result<Vec<Replicated<B>>, BoxError> {
+        let lambda = f.len();
+        let mul = zip(repeat(ctx), a).enumerate().map(|(i, (ctx, &a_bit))| {
+            let f_bit = f[i / lambda];
+            async move { ctx.multiply(RecordId::from(i)).execute(f_bit, a_bit).await }
+        });
+        try_join_all(mul).await
+    }
+
+    /// Step 6. `for j=1..λ, [c_j] = Σ [g_(i,j)] where i=1..λ`
+    ///
+    /// The interpretation is, `c` is formed by taking the "inner product" of `f` and `a`.
+    ///
+    /// # Example
+    /// ```ignore
+    ///   [g] = 0 0 0 0  0 0 1 0  0 0 0 0  0 0 0 0
+    ///   [c] = 0 0 1 0
+    /// ```
+    fn step6(g: &[Replicated<B>], lambda: usize) -> Vec<Replicated<B>> {
+        (0..lambda)
+            .map(|j| {
+                let mut v = Replicated::new(B::ZERO, B::ZERO);
+                (0..g.len()).step_by(lambda).for_each(|i| {
+                    v += g[i + j];
+                });
+                v
+            })
+            .collect()
+    }
+
+    /// Step 7. `for j=1..λ, [b_(,j)] = ∨ [c_k] where k=1..j`
+    ///
+    /// The interpretation is, `b_(,1)..b_(,λ)` are the prefix-or bits of `c`.
+    ///
+    /// # Example
+    /// ```ignore
+    ///   [c] = 0 0 1 0
+    ///   [b] = 0 0 1 1
+    /// ```
+    async fn step7(
+        c: &[Replicated<B>],
+        ctx: ProtocolContext<'_, B>,
+    ) -> Result<Vec<Replicated<B>>, BoxError> {
+        let lambda = c.len();
+        let mut block_or = Vec::with_capacity(lambda);
+        (0..lambda).for_each(|j| block_or.push(Self::block_or(&c[0..=j], j, ctx.clone())));
+        try_join_all(block_or).await
+    }
+
+    /// Step 8. `for i,j=1..λ, [s_(i,j)] = MULT([f_i], [b_(,j)])`
+    ///
+    /// The interpretation is, `s_(i,j)` forms an all-0 vector, except that the i'th block equals `c` iff `f_i` = 1.
+    ///
+    /// # Example
+    /// ```ignore
+    ///   [f] = 0 1 0 0
+    ///   [b] = 0 0 1 1
+    ///   [s] = 0 0 0 0  0 0 1 1  0 0 0 0  0 0 0 0
+    /// ```
+    async fn step8(
+        f: &[Replicated<B>],
+        b: &[Replicated<B>],
+        ctx: ProtocolContext<'_, B>,
+    ) -> Result<Vec<Replicated<B>>, BoxError> {
+        let lambda = f.len();
+        let mut mul = Vec::new();
+        for (i, &f_bit) in f.iter().enumerate().take(lambda) {
+            for (j, &b_bit) in b.iter().enumerate().take(lambda) {
+                let c = ctx.clone();
+                mul.push(
+                    c.multiply(RecordId::from(lambda * i + j))
```

Comment:
Here I *think* you want to keep the record_id that was passed in, and instead use the context to give yourself different randomness for each step.